x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/assignment/TrainedModelAssignment.java

```diff
@@ -153,6 +156,43 @@ public String[] getStartedNodes() {
             .toArray(String[]::new);
     }
 
+    public Optional<String> selectRandomStartedNodeWeighedOnAllocations() {
+        List<String> nodeIds = new ArrayList<>(nodeRoutingTable.size());
+        List<Integer> cumulativeAllocations = new ArrayList<>(nodeRoutingTable.size());
+        int allocationSum = 0;
+        for (Map.Entry<String, RoutingInfo> routingEntry : nodeRoutingTable.entrySet()) {
+            if (RoutingState.STARTED.equals(routingEntry.getValue().getState())) {
+                nodeIds.add(routingEntry.getKey());
+                allocationSum += routingEntry.getValue().getCurrentAllocations();
+                cumulativeAllocations.add(allocationSum);
+            }
+        }
+
+        if (allocationSum == 0) {
+            return Optional.empty();
+        }
+
+        int randomInt = Randomness.get().ints(1, 1, allocationSum + 1).iterator().nextInt();
+        int nodeIndex = Collections.binarySearch(cumulativeAllocations, randomInt);
+        if (nodeIndex < 0) {
+            nodeIndex = -nodeIndex - 1;
+        }
+        return Optional.of(nodeIds.get(nodeIndex));
+    }
+
+    public Optional<String> selectRandomStartedNodeVersionedBefore(
```

Comment:
This is the fallback method used if `selectRandomStartedNodeWeighedOnAllocations` returns `empty` isn't it and the purpose is to pick a node for the inference request?

Why do the version filter when any node in the `STARTED` state could be used in the inference request

Reply:
If a node is upgraded and it had an assignment, when the node is brought back we do not change its state from `STARTED` to `STARTING`, or remove its entry from the routing table. We do not want to run the balancer while the cluster hasn't been fully upgraded. Thus, the model won't be loaded until the full cluster has been upgraded. This is a scenario that only affects mixed clusters from before 8.4 and after 8.4+. We could try to change the state to `STARTING` but even doing that would mean there'll be a window of time when the state will be `STARTED` and we could end up routing a request to that node and subsequently get a failure. A bit tricky but this is one of the easiest fixes to having inference requests routed to nodes running the model in a mixed cluster situation.